funsor/minipyro.py

```diff
@@ -0,0 +1,358 @@
+            inputs[frame.name] = funsor.bint(size)
+    value = funsor.torch.Tensor(data, inputs, output.dtype)
+    assert value.output == output
+    return value
+
+
+# The log_joint messenger is the main way of recording log probabilities.
+# This is roughly the Funsor equivalent to pyro.poutine.trace.
+class log_joint(Messenger):
+    def __enter__(self):
+        super(log_joint, self).__enter__()
+        self.log_factors = OrderedDict()  # maps site name to log_prob factor
+        self.plates = set()
+        return self
+
+    def process_message(self, msg):
+        if msg["type"] == "sample":
+            if msg["value"] is None:
+                # Create a delayed sample.
+                msg["value"] = funsor.Variable(msg["name"], msg["fn"].inputs["value"])
+            elif not isinstance(msg["value"], funsor.Funsor):
+                # Convert Tensors to Funsors.
+                output = msg["fn"].inputs["value"]
+                msg["value"] = tensor_to_funsor(msg["value"], msg["cond_indep_stack"], output)
+
+    def postprocess_message(self, msg):
+        if msg["type"] == "sample":
+            assert msg["name"] not in self.log_factors, "all sites must have unique names"
+            log_prob = msg["fn"](value=msg["value"])
+            # Broadcast, similar to torch.distributions.Distribution.expand().
+            for frame in msg["cond_indep_stack"].values():
+                if frame.name not in log_prob.inputs:
+                    inputs = OrderedDict([(frame.name, funsor.bint(frame.size))])
+                    log_prob += funsor.torch.Tensor(torch.zeros(frame.size), inputs)
+            self.log_factors[msg["name"]] = log_prob
+            self.plates.update(f.name for f in msg["cond_indep_stack"].values())
+
+
+# apply_stack is called by pyro.sample and pyro.param.
+# It is responsible for applying each Messenger to each effectful operation.
+def apply_stack(msg):
+    for pointer, handler in enumerate(reversed(PYRO_STACK)):
+        handler.process_message(msg)
+        # When a Messenger sets the "stop" field of a message,
+        # it prevents any Messengers above it on the stack from being applied.
+        if msg.get("stop"):
+            break
+    if msg["value"] is None:
+        msg["value"] = msg["fn"](*msg["args"])
+
+    # A Messenger that sets msg["stop"] == True also prevents application
+    # of postprocess_message by Messengers above it on the stack
+    # via the pointer variable from the process_message loop
+    for handler in PYRO_STACK[-pointer-1:]:
+        handler.postprocess_message(msg)
+    return msg
+
+
+# sample is an effectful version of Distribution.sample(...)
+# When any effect handlers are active, it constructs an initial message and calls apply_stack.
+def sample(name, fn, obs=None):
+
+    # if there are no active Messengers, we just draw a sample and return it as expected:
+    if not PYRO_STACK:
+        raise NotImplementedError('Funsor cannot sample')
```

Comment:
Can we return `fn(value=Variable(name, ...)).sample({name}).point` here?

Reply:
Let me try to do that...

Reply:
We should probably also check `if not PYRO_STACK and funsor.interpreter._INTERPRETATION is eager` before doing that